ext/mbstring/mbstring.c

```diff
@@ -435,55 +441,25 @@ static bool php_mb_zend_encoding_lexer_compatibility_checker(const zend_encoding
 
 static const zend_encoding *php_mb_zend_encoding_detector(const unsigned char *arg_string, size_t arg_length, const zend_encoding **list, size_t list_size)
 {
-	mbfl_string string;
-
 	if (!list) {
-		list = (const zend_encoding **)MBSTRG(current_detect_order_list);
+		list = (const zend_encoding**)MBSTRG(current_detect_order_list);
 		list_size = MBSTRG(current_detect_order_list_size);
 	}
 
-	mbfl_string_init(&string);
-	string.val = (unsigned char *)arg_string;
-	string.len = arg_length;
-	return (const zend_encoding *) mbfl_identify_encoding(&string, (const mbfl_encoding **)list, list_size, 0);
+	return (const zend_encoding*)mb_guess_encoding((unsigned char*)arg_string, arg_length, (const mbfl_encoding **)list, list_size, false);
 }
 
 static size_t php_mb_zend_encoding_converter(unsigned char **to, size_t *to_length, const unsigned char *from, size_t from_length, const zend_encoding *encoding_to, const zend_encoding *encoding_from)
 {
-	mbfl_string string, result;
-	mbfl_buffer_converter *convd;
-
-	/* new encoding */
-	/* initialize string */
-	string.encoding = (const mbfl_encoding*)encoding_from;
-	string.val = (unsigned char*)from;
-	string.len = from_length;
-
-	/* initialize converter */
-	convd = mbfl_buffer_converter_new((const mbfl_encoding *)encoding_from, (const mbfl_encoding *)encoding_to, string.len);
-	if (convd == NULL) {
-		return (size_t) -1;
-	}
-
-	mbfl_buffer_converter_illegal_mode(convd, MBSTRG(current_filter_illegal_mode));
-	mbfl_buffer_converter_illegal_substchar(convd, MBSTRG(current_filter_illegal_substchar));
-
-	/* do it */
-	size_t loc = mbfl_buffer_converter_feed(convd, &string);
-
-	mbfl_buffer_converter_flush(convd);
-	mbfl_string_init(&result);
-	if (!mbfl_buffer_converter_result(convd, &result)) {
-		mbfl_buffer_converter_delete(convd);
-		return (size_t)-1;
-	}
-
-	*to = result.val;
-	*to_length = result.len;
+	unsigned int num_errors = 0;
+	zend_string *result = mb_fast_convert((unsigned char*)from, from_length, (const mbfl_encoding*)encoding_from, (const mbfl_encoding*)encoding_to, MBSTRG(current_filter_illegal_substchar), MBSTRG(current_filter_illegal_mode), &num_errors);
 
-	mbfl_buffer_converter_delete(convd);
+	*to_length = ZSTR_LEN(result);
+	*to = emalloc(ZSTR_LEN(result) + 1); /* Include terminating null byte */
+	memcpy(*to, ZSTR_VAL(result), ZSTR_LEN(result) + 1);
+	zend_string_free(result);
 
-	return loc;
+	return from_length;
 }
```

Comment:
Just staring again at this why are we returning ``from_length`` when this is an input parameter?

Reply:
Note that the function is declared to return `size_t`. If you look at the legacy code, it would process bytes until the conversion was aborted due to an error (which never happened in practice), then return the number of bytes processed.

In this case, the "number of bytes processed" is *always* the length of the string, so we return that.

In short, there is actually no point in returning the number of bytes processed, but that's what the function is expected to do.

Reply:
We could change the function signature, I guess... it will mean also changing the call sites in the Zend engine.

Reply:
Let's maybe do that in a separate PR then :), But thanks for the explanation!